hibernate-core/src/main/java/org/hibernate/dialect/pagination/Oracle12LimitHandler.java

```diff
@@ -0,0 +1,170 @@
+
+		final boolean hasFirstRow = LimitHelper.hasFirstRow( selection );
+		final boolean hasMaxRows = LimitHelper.hasMaxRows( selection );
+
+		if ( !hasMaxRows ) {
+			return sql.trim();
+		}
+
+		final LockOptions lockOptions = queryParameters.getLockOptions();
+		if ( lockOptions != null ) {
+			final LockMode lockMode = lockOptions.getLockMode();
+			switch ( lockMode ) {
+				case UPGRADE:
+				case PESSIMISTIC_READ:
+				case PESSIMISTIC_WRITE:
+				case UPGRADE_NOWAIT:
+				case FORCE:
+				case PESSIMISTIC_FORCE_INCREMENT:
+				case UPGRADE_SKIPLOCKED:
+					return processSql( sql, selection );
+				default:
+					return processSqlOffsetFetch( sql, hasFirstRow );
+			}
+		}
+		return processSqlOffsetFetch( sql, hasFirstRow );
+	}
+
+	private String processSqlOffsetFetch(String sql, boolean hasFirstRow) {
+		bindLimitParametersInReverseOrder = false;
+		final int forUpdateLastIndex = getForUpdateIndex( sql );
+
+		if ( forUpdateLastIndex > -1 ) {
+			return processSql( sql, forUpdateLastIndex, hasFirstRow );
+		}
+
+		sql = sql.trim();
+		final int offsetFetchLength;
+		final String offsetFetchString;
+		if ( hasFirstRow ) {
+			offsetFetchString = " offset ? rows fetch next ? rows only";
+		}
+		else {
+			offsetFetchString = " fetch first ? rows only";
+		}
+		offsetFetchLength = sql.length() + offsetFetchString.length();
+
+		return new StringBuilder( offsetFetchLength ).append( sql ).append( offsetFetchString ).toString();
+	}
+
+	private String processSql(String sql, int forUpdateIndex, boolean hasFirstRow) {
+		bindLimitParametersInReverseOrder = true;
+		sql = sql.trim();
+
+		String forUpdateClause = null;
+		boolean isForUpdate = false;
+		if ( forUpdateIndex > -1 ) {
+			// save 'for update ...' and then remove it
+			forUpdateClause = sql.substring( forUpdateIndex );
+			sql = sql.substring( 0, forUpdateIndex - 1 );
+			isForUpdate = true;
+		}
+
+		final StringBuilder pagingSelect;
+
+		final int forUpdateClauseLength;
+		if ( forUpdateClause == null ) {
+			forUpdateClauseLength = 0;
+		}
+		else {
+			forUpdateClauseLength = forUpdateClause.length() + 1;
+		}
+
+		if ( hasFirstRow ) {
+			pagingSelect = new StringBuilder( sql.length() + forUpdateClauseLength + 98 );
+			pagingSelect.append( "select * from ( select row_.*, rownum rownum_ from ( " );
+			pagingSelect.append( sql );
+			pagingSelect.append( " ) row_ where rownum <= ?) where rownum_ > ?" );
+		}
+		else {
+			pagingSelect = new StringBuilder( sql.length() + forUpdateClauseLength + 37 );
+			pagingSelect.append( "select * from ( " );
+			pagingSelect.append( sql );
+			pagingSelect.append( " ) where rownum <= ?" );
+		}
+
+		if ( isForUpdate ) {
+			pagingSelect.append( " " );
+			pagingSelect.append( forUpdateClause );
+		}
+
+		return pagingSelect.toString();
+	}
+
+	private int getForUpdateIndex(String sql) {
+		final int forUpdateLastIndex = sql.toLowerCase( Locale.ROOT ).lastIndexOf( "for update" );
```

Comment:
Is there such possibility that the sql string was provided by user and it happens to have two empty chars between 'for' and 'update' (or it doesn't use single quote but double quote)?

Reply:
I think the point related to the spaces between `for` and `update` is true while the single and double quote I think it does not matter here

Reply:
I simply gave some example regarding the robustness concern of such SQL string simple tweaking. The possibilities might be much more than we can expect now and ANTLR or semantic clue might be the only reliable way to be fool-proof. Yeah, we have to process a SQL string here, but in v6 we might face a semantic representation and then we can solve the issue in more elegant way.